### ISSUE TITLE
bug(ctm): _apply_proj_unfused is 0.83 relative wrong vs dense on the default charged U(1)-Sz path (mixed-charge projector defeats output-target inference)

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#871]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

`_apply_proj_unfused` returns a tensor **0.83 relative** away from the same
contraction with its operands densified, on the **default** CTM path, for a
charged U(1)-Sz state. No error, no warning. This is #834's mechanism reaching
production, at a site that had not been measured.

It is filed separately from #834 because the mechanism is different: not leg
pairing, but **output-target inference over a mixed-charge operand**.

## Measurement

Every `_contract_symmetric` call of a D=2, χ=8 charged U(1)-Sz sweep
(`heisenberg_u1sz_init_pair`, `ctm_tensor_2site`, `max_iter=3`), scored against
`contract_with_subscripts` on the **densified same operands**. Dense is ground
truth: a plain einsum has no block structure to mis-pair.

| call site | calls | max relative gap |
|---|---|---|
| **`_ctm_tensor_moves.py:136 _apply_proj_unfused`** | **56** | **8.260e-01** |
| `_ctm_tensor_moves.py:136 _apply_proj_unfused` (other 40) | 40 | 1.7e-16 |
| `_ctm_tensor_projector_2x2.py _build_enlarged_corner` (6 frames) | 106 | 0.0 |
| `_ctm_tensor_projector_2x2.py _compute_2x2_projector_symmetric` (8 frames) | 96 | ≤5.7e-17 |
| `_ctm_tensor_moves.py _ctm_tensor_absorb_*_2plaq` (4) | 16 | 0.0 |
| `_ctm_tensor_init.py:98 _build_double_layer_tensor` | 2 | 0.0 |

48 of the 96 `_apply_proj_unfused` calls are wrong. Everything else in the sweep
is exact to rounding.

## Mechanism

`_parse_contraction_prelude` infers the output target by fusing the operands'
net charges, but **only credits a tensor whose blocks all agree**:

```python
targets = np.unique(_net_charges(tensor.indices, tensor._block_keys))
if targets.size == 1:
    total = sym.fuse(total, targets)
```

The unfused projector is **mixed-charge** — its blocks span net charges
`[-2, 0, 2]` — so it contributes the identity, `output_target` comes out `None`,
and every product landing on ±2 fails `output_key not in valid_output_set` and is
dropped by the bare `continue`.

Correlation over the 96 calls, printing operand net charges and the inferred
target:

```
--- WRONG ---
  gap=3.548e-01  dfa,defbc->aebc   target=None  nets=[[-2, 0, 2], [0]]
  gap=2.210e-01  aecd,bec->adb     target=None  nets=[[0], [-2, 0, 2]]
  gap=1.300e-01  aec,efbcd->afbd   target=None  nets=[[-2, 0, 2], [0]]
  gap=3.526e-01  aecd,edb->acb     target=None  nets=[[0], [-2, 0, 2]]
--- exact ---
  gap=0.000e+00  bad,adc->bc       target=None  nets=[[-2, 0, 2], [0]]
  gap=0.000e+00  acb,acd->bd       target=None  nets=[[-2, 0, 2], [0]]
```

A mixed-charge operand is **necessary but not sufficient**: it is present in the
exact calls too, and what separates them is whether any surviving product
actually lands on a non-identity target. So the discard is the proximate cause;
the mixed-charge projector is what makes it reachable.

Independently, counting discarded products and evaluating each one:

| site | products discarded | non-zero | max \|product\| |
|---|---|---|---|
| `_apply_proj_unfused` | 560 | **192** | **8.15** |
| `_build_enlarged_corner` (6 frames) | ~2000 | 0 | — |

The enlarged-corner row is the case #852 documents: it discards constantly and
is exact anyway because the rank-1 corners have no weight in those sectors. This
site is not that.

## Why it was not caught

- #852 instrumented `_build_enlarged_corner` only, and correctly found it exact.
  The [correction on #834](https://github.com/tenax-lab/tenax/issues/834#issuecomment-5249440831)
  generalised that to "the default path is clean". **That generalisation is
  wrong** and I am retracting it; the site next door is not clean.
- The #605 work that introduced `_apply_proj_unfused` verified it byte-identical
  to the fuse-then-contract path it replaced. Both discard the same products, so
  agreeing with each other could not detect this.

## Open questions, deliberately not answered here

- **Is the projector supposed to be mixed-charge?** Net charges spanning
  `[-2, 0, 2]` mean it is not a conserving tensor. If that is the real defect,
  the fix belongs upstream in `_compute_2x2_projector_symmetric` rather than in
  target inference.
- **What does it cost physically?** Not measured. The environment is
  gauge-dependent, so a wrong intermediate does not automatically move the
  energy. The right experiment is a symmetric-vs-dense ground-state energy on the
  same charged D=2 state.

## Reproduction

`TENAX_STRICT_CONTRACT=1` (added in the #834 change) turns the discard into an
error, so the sweep refuses instead of silently continuing:

```python
import jax
from tenax.algorithms._ctm_tensor import ctm_tensor_2site
from tenax.algorithms.ipeps import heisenberg_u1sz_init_pair

A, B = heisenberg_u1sz_init_pair(D=2, key=jax.random.PRNGKey(0))
ctm_tensor_2site(A, B, chi=8, max_iter=3, conv_tol=1e-8)   # ValueError under the flag
```

Pinned by `test_strict_mode_reports_real_weight_loss_in_the_ctm_sweep` in
`tests/test_contract_leg_pairing_834.py`, which fails when this is fixed.

Refs #834, #762, #605, #852.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Closing as fixed by #905 / PR #923 (3e40d77). Root cause was an unconditional flow-flip in `_apply_proj_unfused` that left the projector with a nonzero net charge (2*q_chi_new), annihilating every charged U(1)-Sz sector. The current `_apply_proj_unfused` (src/tenax/algorithms/_ctm_tensor_moves.py:140) does a bare split_index + relabel + contract with no flip. Guarded by `test_strict_mode_finds_no_weight_loss_in_the_ctm_sweep` (tests/test_contract_leg_pairing_834.py:556), which runs the charged sweep under TENAX_STRICT_CONTRACT=1 and asserts no weight loss — converted from the original negative pin to a positive one.
